core/src/main/scala/cats/data/NonEmptyList.scala

```diff
@@ -0,0 +1,193 @@
+package cats
+package data
+
+import cats.instances.list._
+import cats.syntax.eq._
+
+import scala.annotation.tailrec
+import scala.collection.mutable.ListBuffer
+
+/**
+ * A data type which represents a non empty list of A, with
+ * single element (head) and optional structure (tail).
+ */
+final case class NonEmptyList[A](head: A, tail: List[A]) {
+
+  /**
+   * Return the head and tail into a single list
+   */
+  def toList: List[A] = head :: tail
+
+  /**
+   *  Applies f to all the elements of the structure
+   */
+  def map[B](f: A => B): NonEmptyList[B] =
```

Comment:
One thing that we could do would be to define it as a `sealed abstract class` so we have some flexibility in changing it in the future. However, we still wouldn't be able to change it into a value class without breaking binary compatibility.